File: Algorithms/0090_Subsets_II/Java/SubsetsIISolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    List<List<Integer>> res = new ArrayList<>();
    boolean[] status = null;
    Set<String> cache = new HashSet<>();

    public List<List<Integer>> subsetsWithDup(int[] nums) {

        if (nums.length == 0) return new ArrayList<>();
        status = new boolean[nums.length];
        Arrays.fill(status, false);
        Arrays.sort(nums);
        for (int i = 0; i < nums.length; i++) {
            generateSubset(nums, i, nums.length, new ArrayList<>());
        }
        return res;
    }

    private void generateSubset(int[] nums, int index, int limit, ArrayList<Integer> temp) {
        if (temp.size() > limit) return;
        if (!cache.contains(temp.toString())) {
            res.add(new ArrayList<>(temp));
            cache.add(temp.toString());
        }

        for (int i = index; i < nums.length; i++) {
            if (!status[i]) {
                temp.add(nums[i]);
                status[i] = true;
                generateSubset(nums, i + 1, limit, temp);
                status[i] = false;
                temp.remove(temp.size() - 1);
            }
        }
    }
}

File: Algorithms/0090_Subsets_II/Python/Subsets_II_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def subsetsWithDup(self, nums):
        length = len(nums)
        if length == 0: return [[]]
        if length == 1: return [[], nums]
        nums = sorted(nums)
        self.res = []
        self.status = [False for _ in range(length)]
        for i in range(length):
            self.generate_subset(nums, length, [], i)
        return self.res

    def generate_subset(self, alist, limit, temp, start_index):
        if limit < len(temp): return
        if temp not in self.res: self.res.append(temp[:])

        for i in range(start_index, len(alist)):
            if not self.status[i]:
                temp.append(alist[i])
                self.status[i] = True
                self.generate_subset(alist, limit, temp, i + 1)
                self.status[i] = False
                temp.pop()




